### PR TITLE
PW gift card implementaion

### PR DESCRIPTION
**Pull request by poojarazor**

### CODE REVIEW

File: includes/api/shipping-info.php [hunk 1]
```diff
@@ -233,6 +233,12 @@ function prepareRatesResponse1cc($package, $vendorId, $orderId, $address)
         $response['shipping_fee']     = isset($response[0]['price']) ? $response[0]['price'] : 0;
         $response['shipping_fee_tax'] = !empty($response[0]['taxes']) ? 0 : 0; //By default tax is considered as zero.
     }
+
+    // check shipping fee for gift card product
+    if(giftCardProduct($orderId)){
+        $response['shipping_fee'] = 0;
+    }
+
```
Comment (ChetanGN) [inline includes/api/shipping-info.php:241]: Also, can you please fix this line 
https://github.com/razorpay/razorpay-woocommerce/blob/d97a12ea126829c9e3194069a9e6859553df01cc/includes/api/shipping-info.php#L294
It's adding one more DB operation to all shipping API calls irrespective of whether a merchant using smart cod or not.

File: includes/api/shipping-info.php [hunk 2]
```diff
@@ -323,6 +329,38 @@ function getCodShippingInfo1cc($instanceId, $methodId, $orderId, $address)
...
+    $items = $order->get_items();
+
+    $giftCount = 0;
+    $cartCount  = 0;
+    foreach ($order->get_items() as $item_id => $item)  {
+        $cartCount++;
+        $product = $item->get_product();
+
+        if($product->is_type('variation')){
+             $parent_product_id = $product->get_parent_id();
```
Comment (ChetanGN) [inline includes/api/shipping-info.php:343]: All variables should be in the same format i.e **camelcase**

File: includes/api/shipping-info.php [hunk 3]
```diff
@@ -323,6 +329,38 @@ function getCodShippingInfo1cc($instanceId, $methodId, $orderId, $address)
...
+             $parent_product = wc_get_product($parent_product_id);
+             
+            if($parent_product->get_type() == 'pw-gift-card' || $parent_product->get_type() == 'gift-card'){
+                $giftCount++;
```
Comment (ChetanGN) [inline includes/api/shipping-info.php:347]: Change the variable name to `giftProductCount`

**Review by ChetanGN — CHANGES_REQUESTED** [review]

File: woo-razorpay.php [hunk 4]
```diff
@@ -1824,6 +1849,140 @@ public function update1ccOrderWC(& $order, $wcOrderId, $razorpayPaymentId)
...
+
+                    $usedAmt = $giftcard['value']/100;
+                    $giftCode = $giftcard['code'];
+                    if(is_plugin_active('yith-woocommerce-gift-cards/init.php')){
+
+                        $yithCard = new YITH_YWGC_Gift_Card( $args = array('gift_card_number'=> $giftCode));
+
+                        $giftCardBalance = $yithCard->get_balance();
+
+                        if($giftCardBalance == null && $balance >= 0 && $usedAmt > $giftCardBalance ){
```
Comment (ChetanGN) [inline woo-razorpay.php:1869]: What is `$balance`?

File: woo-razorpay.php [hunk 5]
```diff
@@ -1824,6 +1849,140 @@ public function update1ccOrderWC(& $order, $wcOrderId, $razorpayPaymentId)
...
+                                    $orderId
+                                )
+                            );
+
+                            $itemId = $wpdb->insert_id;
+                            wc_add_order_item_meta( $itemId, '_line_total', -$usedAmt );
+                        }
+
+                    }
+                    else if(is_plugin_active('pw-woocommerce-gift-cards/pw-gift-cards.php'))
```
Comment (ChetanGN) [inline woo-razorpay.php:1893]: What if the merchant is installed and activated both GC plugins? How this logic react?

File: woo-razorpay.php [hunk 6]
```diff
@@ -1824,6 +1849,140 @@ public function update1ccOrderWC(& $order, $wcOrderId, $razorpayPaymentId)
...
+                        $order->save();
+
+                        rzpLogInfo("Coupon details updated for orderId: $orderId");
+                    }
+                }
+                
+            }
+        }
+
+        public function debitGiftCards( $orderId, $order, $note, $usedAmt, $giftCardNo) {
```
Comment (ChetanGN) [inline woo-razorpay.php:1947]: If this plugin only used in this class, make it protected.

File: woo-razorpay.php [hunk 7]
```diff
@@ -1824,6 +1849,140 @@ public function update1ccOrderWC(& $order, $wcOrderId, $razorpayPaymentId)
...
+                            }else{
+                                //Deduct amount of gift card
+                               $this->debitGiftCards($orderId, $order, "order_id: $orderId checkout_update_order_meta", $usedAmt);
+                            }
+
+                        }else{
+                            $this->process_refund($orderId, $razorpayData['amount_paid'], $reason = '', $razorpayPaymentId);
+                        }
+
+                    }else{
```
Comment (ChetanGN) [inline woo-razorpay.php:1911]: If GC is applied then the customer can't apply coupons?

**Review by ChetanGN — CHANGES_REQUESTED** [review]

File: includes/api/shipping-info.php (hunk 1 above)
Comment (poojarazor) [reply includes/api/shipping-info.php:241]: it is not specific for smart COD , we are using $order->get_total() To verify the min order amount

File: woo-razorpay.php (hunk 7 above)
Comment (poojarazor) [reply woo-razorpay.php:1911]: Customer can use both at a time

File: woo-razorpay.php (hunk 6 above)
Comment (poojarazor) [reply woo-razorpay.php:1947]: ok

File: woo-razorpay.php (hunk 5 above)
Comment (poojarazor) [reply woo-razorpay.php:1893]: Customer can use only one plugin at a time

File: woo-razorpay.php [hunk 8]
```diff
@@ -1194,21 +1209,19 @@ protected function getOrderKey($order)
...
 
-            if (! $order or ! $order->get_transaction_id())
+            if (! $order or ! $razorpayPaymentId)
             {
                 return new WP_Error('error', __('Refund failed: No transaction ID', 'woocommerce'));
             }
 
             $client = $this->getRazorpayApiInstance();
 
-            $paymentId = $order->get_transaction_id();
```
[the remark this answers is not in the document]
Comment (poojarazor) [reply woo-razorpay.php:1208]: Transaction id is not updated yet

**Review by angbpy — APPROVED** [review]

**Review by ChetanGN — APPROVED** [review]

**Pull request merged**